pulsar-client-cpp/lib/KeyValueImpl.cc

```diff
@@ -0,0 +1,78 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <pulsar/Schema.h>
+#include "SharedBuffer.h"
+#include "KeyValueImpl.h"
+
+using namespace pulsar;
+
+namespace pulsar {
+
+KeyValueImpl::KeyValueImpl() {}
+
+KeyValueImpl::KeyValueImpl(const char *data, int length, const KeyValueEncodingType &keyValueEncodingType) {
+    keyValueEncodingType_ = keyValueEncodingType;
+    if (keyValueEncodingType == KeyValueEncodingType::INLINE) {
+        SharedBuffer buffer = SharedBuffer::wrap(const_cast<char *>(data), length);
+        int keySize = buffer.readUnsignedInt();
+        if (keySize != -1) {
+            SharedBuffer keyContent = buffer.slice(0, keySize);
+            keyContent_ = std::string(keyContent.data(), keySize);
+            buffer.consume(keySize);
+        }
+
+        int valueSize = buffer.readUnsignedInt();
+        if (valueSize != -1) {
+            SharedBuffer valueContent = buffer.slice(0, valueSize);
+            valueContent_ = std::string(valueContent.data(), valueSize);
```

Comment:
It's better to avoid the copy in the internal implementation. We should save `valueContent` (and the `keyContent` in line 35) as the field of `KeyValueImpl`. See how `Message` does.

You can also see the original PR for how to optimize the data copy: https://github.com/apache/pulsar/pull/13293